packages/errors/src/error.ts

```diff
@@ -0,0 +1,37 @@
+import { SolanaErrorCode } from './codes';
+import { SolanaErrorContext } from './context';
+import { getErrorMessage } from './message-formatter';
+
+const SOLANA_ERROR_TAG = Symbol();
+
+export function isSolanaError<TErrorCode extends SolanaErrorCode>(
+    e: unknown,
+    code?: TErrorCode,
+): e is SolanaError<TErrorCode> {
+    const isSolanaError = !!e && typeof e === 'object' && SOLANA_ERROR_TAG in e;
+    if (isSolanaError) {
+        if (code !== undefined) {
+            return (e as SolanaError<TErrorCode>).context.__code === code;
+        }
+        return true;
+    }
+    return false;
+}
+
+type SolanaErrorConfig = Readonly<{
+    [P in SolanaErrorCode]: Readonly<
+        (SolanaErrorContext[P] extends undefined ? object : SolanaErrorContext[P]) & { __code: P }
+    >;
+}>;
+
+export class SolanaError<TErrorCode extends SolanaErrorCode = SolanaErrorCode> extends Error {
+    // This gives us a way for the `isSolanaError()` function to identify a `SolanaError` without
+    // having to import the class for use in an `instanceof` check.
+    readonly [SOLANA_ERROR_TAG] = 1;
```

Comment:
Aren't global symbols just as susceptible to dual package hazards as `instanceof` checks?

An alternative would be to use something like `readonly __solanaError = 1` and then have an opaque type with a TypeScript `unique symbol` to ensure it's coming from this specific class.

Reply:
Interesting.

The motivation behind this was never to avoid dual package hazards, it was to make it so that you could import `isSolanaError()` without taking a dependency on `class SolanaError`.

Alternatives that eliminate the dual-package hazard, but open the possibility of ‘spoofing’ some other object as a `SolanaError`:

* Instead of a unique symbol as I've used here, use the global `Symbol.for('sOlAnAeRrOr')` or whatever that can be recovered across packages
* Even easier, just check `e.name === 'SolanaError'`

God I hate classes.

Reply:
I see you went with `e.name === 'SolanaError'` which I quite like tbh. 👌 

Is it possible to add a TypeScript opaque attribute to a class so, at least, TypeScript prevents the spoofing of the class?

Reply:
> Is it possible to add a TypeScript opaque attribute…

It wont be useful for any purpose, because JS errors thrown [are always](https://www.typescriptlang.org/play?#code/C4JwngBA3gUBEHoBUEB06JITAvhAxgIbD4AWEAFAKYCU0c8jCCEAegPy5A) `unknown`.

```ts
try {
  /* ... */
} catch (e) {
      // ^? e: unknown
}
```